keps/sig-testing/5468-invariant-testing/README.md

```diff
@@ -0,0 +1,960 @@
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Enable end to end tests to check invariants after the rest of the test suite
+- Prevent unowned tests
+- Allow opting into / out-of this testing
+- Prevent flaking, unreliable tests
+- Ensure result reporting is structured
+- Must not impact the conformance test suite
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Enable completely arbitrary checks
+- Targeting integration tests.
+  - We are specifically aiming for end to end tests for this purpose.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+First, all invariant **MUST** have clearly documented owners, which will be surfaced
+with the failure results.
+No invariant tests may roll up to the general test owners.
+There is precedent in documenting featuregate and metrics owners.
+
+A shared system will be introduced to the e2e framework to enable this form of
+testing.
+
+All utilities for this will be owned by the SIG Testing Leads and thoroughly
+vetted for reliability.
+
+New invariants will be tested many times before merge, and may be demoted
+at any time by tagging the associated sentinel test `[Flaky]`. Failure to respond
+to flaking invariant tests in a timeline fashion will result in demoting or
+removing them.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+If implemented poorly, this could result in tests flaking in any number of e2e
+test CI jobs that are now running these tests.
+
+We will mitigate this by thoroughly reviewing and testing the invariant checking utilities, 
+and by limiting to only portable, reliable mechanisms.
+
+If not run in many CI jobs, there will be limited benefit to the signal.
+We will aim to generally introduce these as default selected tests.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+[ginkgo], which we use for the e2e tests, does not have a facility for 
```

Comment:
Nit: `Ginkgo` is a name and always spelled with a capital letter. Applies throughout the text.